File: leetcode/problems/problem-specific/46.permutations/46.permutations.go
package main

func permute(nums []int) [][]int {
	if len(nums) <= 0 {
		return [][]int{}
	}

	res := [][]int{}
	permuteHelper(nums, make([]bool, len(nums)), []int{}, &res)
	return res
}

func permuteHelper(nums []int, used []bool, permutation []int, res *[][]int) {
	if len(permutation) == len(nums) {
		*res = append(*res, append([]int{}, permutation...))
		return
	}

	for i := range nums {
		if used[i] {
			continue
		}

		used[i] = true
		permutation = append(permutation, nums[i])
		permuteHelper(nums, used, permutation, res)
		permutation = permutation[:len(permutation)-1]
		used[i] = false
	}
}

File: leetcode/problems/problem-specific/46.permutations/46.permutations.py
#
# @lc app=leetcode id=46 lang=python3
#
# [46] Permutations
#

# @lc code=start

# * Backtracking Solution | O(n*(n!)) Time | O(n) Space
# * n -> The length of nums array


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        def permuteHelper(used, permutation):
            if len(permutation) == len(nums):
                res.append(permutation[:])
                return

            for num in nums:
                if num in used:
                    continue

                used.add(num)
                permutation.append(num)
                permuteHelper(used, permutation)
                permutation.pop()
                used.remove(num)

        permuteHelper(set(), [])
        return res


# @lc code=end
